File: ts/webhook-processor/ingest/topic.ts
import { Topic } from "encore.dev/pubsub";

export interface WebhookEvent {
  source: string;
  event_type: string;
  payload: Record<string, unknown>;
  received_at: string;
}

// Topic for incoming webhook events. Multiple services can subscribe
// to process events independently (fan-out pattern).
export const WebhookTopic = new Topic<WebhookEvent>("webhook-received", {
  deliveryGuarantee: "at-least-once",
});

File: webhook-processor/ingest/topic.go
package ingest

import (
	"encoding/json"

	"encore.dev/pubsub"
)

// WebhookEvent represents an incoming webhook event.
type WebhookEvent struct {
	Source     string          `json:"source"`
	EventType string          `json:"event_type"`
	Payload   json.RawMessage `json:"payload"`
	ReceivedAt string         `json:"received_at"`
}

// WebhookTopic is the Pub/Sub topic for incoming webhook events.
// Multiple services can subscribe to process events independently (fan-out pattern).
var WebhookTopic = pubsub.NewTopic[*WebhookEvent]("webhook-received", pubsub.TopicConfig{
	DeliveryGuarantee: pubsub.AtLeastOnce,
})
